### DISCUSSION TITLE
We are now consistent on Rounding, finally!!!

### DISCUSSION THREAD

**Discussion by gcarreno** [ObjectPascal-Community/1brc-ObjectPascal#71] [Announcements]
Hey Y'All,

Looks like there is a simple fix: Not using Double but Currency.

```pas
function RoundExDouble(x: Currency): Double;
begin
  Result := Ceil(x * 10) / 10;
end;
```

Does not fix the issue that Delphi is not consistent with Double across Windows 32,64 and Linux 64.

But at least we now have consistency on our end!!!

Many thanks to @paweld for spotting the fix!!

Cheers,
Gus

**Comment by synopse** [comment]
Something is wrong here.
We should not need to use Currency to have the same results.

And it is worth saying that **Currency is broken on FPC when you cross-compile from Win64 to Win32**.
https://gitlab.com/freepascal.org/fpc/source/-/issues/40133
We should better stay away from currency.

We could try to use our own version and ignore the Math unit, as I do in my entry:
```
function ceil(x: double): integer; // "official" rounding method
begin
  result := trunc(x) + ord(frac(x) > 0);  // using FPU is fast enough here
end;
```

On my PC, it gives consistent results on FPC and Delphi.

**Reply by synopse** [reply]
I prepared a Pull Request about rounding, but since I have a pending PR about my entry, I don't want to mix things.

Just change RoundEx.Common as such:
```
ENDIF}

interface

uses
  SysUtils;

function RoundExDouble(x: double): Double;
function RoundExInteger(x: double): Integer;
function GenerateProgressBar(APBPosition, APBMax, APBWIdth: Integer): String;

implementation

// we define our own "official" cross-compiler ceil() rounding method: the math
// unit seems not consistent between Delphi and FPC

function ceil(x: double): integer;
begin
  result := trunc(x) + ord(frac(x) > 0);  // using FPU is fast enough here
end;

function RoundExDouble(x: double): Double;
begin
  Result := Ceil(x * 10) / 10;
end;

function RoundExInteger(x: double): Integer;
begin
  Result := Ceil(x * 10);
end;

...

**Reply by gcarreno** [reply]
Someone from Delphi spent 3 days pulling his hairs because he had a single line not match the correct value, with a difference of `0.1`
I mentioned it and @paweld solved the inconsistency with _Currency_.
Of course, the code presented on the Rounding Section is not an imposition, but a suggestion.
As long as anyone's code can mimic those outputs, with whatever code, then it's all good.

**Reply by synopse** [reply]
He should have better looked at my code, then, or asked here in the discussion and kept his hairs. :haircut: 

The problem comes from the overloads in the Delphi math unit.

Another workaround, still using math, is to make the computation in two steps:
```
function RoundExDouble(x: double): Double;
begin
  x := x * 10;
  Result := Ceil(x) / 10;
end;

function RoundExInteger(x: double): Integer;
begin
  x := x * 10;
  Result := Ceil(x);
end;
```
It also gives the right value, by ensuring that the appropriate `Ceil(x: double)` overload is called.

This is a "classical" problem with Delphi, overloaded functions, and floating points values.

Using `currency`  is an awful workaround, and should not be kept in any serious repository.

**Reply by gcarreno** [reply]
> Just change RoundEx.Common as such:

I'll keep that code in mind. Thanks.
The current version on the `README` is consistent.
Having another round of "Hey peeps, change it again" would not be advisable.
I still need to extract that code and prepare a bug report for Ian Barker.
I still think that Delphi not being coherent on such a basic function across platforms is wrong.

**Reply by gcarreno** [reply]
> He should have better looked at my code, then, or asked here in the discussion and kept his hairs. 💇

The problem is probably mine.
I'm coordinating this challenge in:
- Discord: 2 servers
- Telegram
- Lazarus Forums
- Delphi Praxis Forums
- These discussions
Too many places!!!

But I don't want to exclude anyone, or force them here or anywhere else.

**Reply by synopse** [reply]
I thought we were here to share and progress?
If our code is cleaner, just change it again.
:laughing: 

And you won't have any bug report accepted by Embarcadero.
From the low-level point of view, it "works as expected".

**Comment by synopse** [comment]
It sounds like if my latest PR included my new `ceil()` function within the RoundEx project...

**Reply by gcarreno** [reply]
Yeah, I'll put your code somewhere.
I'll revert the spelunking code at some point. It's not important to the challenge.
